File: Session-3/DDD-Sample-codes/rtgs-return-management-springboot/src/main/java/com/bank/rtgs/returnmanagement/application/service/RtgsReturnApplicationService.java
package com.bank.rtgs.returnmanagement.application.service;

import com.bank.rtgs.returnmanagement.application.dto.CreateReturnRequest;
import com.bank.rtgs.returnmanagement.application.dto.RtgsReturnResponse;
import com.bank.rtgs.returnmanagement.domain.model.ReturnReason;
import com.bank.rtgs.returnmanagement.domain.model.RtgsReturn;
import com.bank.rtgs.returnmanagement.domain.model.TransactionReference;
import com.bank.rtgs.returnmanagement.domain.ports.ReturnPayloadGenerator;
import com.bank.rtgs.returnmanagement.domain.ports.RtgsReturnRepository;
import org.springframework.stereotype.Service;

import java.time.Instant;
import java.util.UUID;

@Service
public class RtgsReturnApplicationService {
    private final RtgsReturnRepository repository;
    private final ReturnPayloadGenerator payloadGenerator;

    public RtgsReturnApplicationService(RtgsReturnRepository repository, ReturnPayloadGenerator payloadGenerator) {
        this.repository = repository;
        this.payloadGenerator = payloadGenerator;
    }

    public RtgsReturnResponse create(CreateReturnRequest request) {
        RtgsReturn rtgsReturn = new RtgsReturn(
                new TransactionReference(request.originalTransactionReference()),
                new ReturnReason(request.reasonCode(), request.reasonDescription()),
                Instant.now()
        );
        repository.save(rtgsReturn);
        return RtgsReturnResponse.from(rtgsReturn);
    }

    public RtgsReturnResponse generatePayload(UUID id) {
        RtgsReturn rtgsReturn = getAggregate(id);
        String payload = payloadGenerator.generate(rtgsReturn);
        rtgsReturn.generatePayload(payload, Instant.now());
        repository.save(rtgsReturn);
        return RtgsReturnResponse.from(rtgsReturn);
    }

    public RtgsReturnResponse markSent(UUID id) {
        RtgsReturn rtgsReturn = getAggregate(id);
        rtgsReturn.markAsSent(Instant.now());
        repository.save(rtgsReturn);
        return RtgsReturnResponse.from(rtgsReturn);
    }

    public RtgsReturnResponse acknowledge(UUID id) {
        RtgsReturn rtgsReturn = getAggregate(id);
        rtgsReturn.acknowledge(Instant.now());
        repository.save(rtgsReturn);
        return RtgsReturnResponse.from(rtgsReturn);
    }

    public RtgsReturnResponse get(UUID id) {
        return RtgsReturnResponse.from(getAggregate(id));
    }

    private RtgsReturn getAggregate(UUID id) {
        return repository.findById(id).orElseThrow(() -> new IllegalArgumentException("RTGS Return not found: " + id));
    }
}

File: Session-3/DDD-Sample-codes/rtgs-return-management-dotnet10/Application/Services/RtgsReturnApplicationService.cs
using Rtgs.ReturnManagement.Api.Application.Dtos;
using Rtgs.ReturnManagement.Api.Domain.Model;
using Rtgs.ReturnManagement.Api.Domain.Ports;

namespace Rtgs.ReturnManagement.Api.Application.Services;

public sealed class RtgsReturnApplicationService
{
    private readonly IRtgsReturnRepository _repository;
    private readonly IReturnPayloadGenerator _payloadGenerator;

    public RtgsReturnApplicationService(IRtgsReturnRepository repository, IReturnPayloadGenerator payloadGenerator)
    {
        _repository = repository;
        _payloadGenerator = payloadGenerator;
    }

    public RtgsReturnResponse Create(CreateReturnRequest request)
    {
        var rtgsReturn = new RtgsReturn(
            new TransactionReference(request.OriginalTransactionReference),
            new ReturnReason(request.ReasonCode, request.ReasonDescription),
            DateTime.UtcNow);

        _repository.Save(rtgsReturn);
        return RtgsReturnResponse.From(rtgsReturn);
    }

    public RtgsReturnResponse GeneratePayload(Guid id)
    {
        var rtgsReturn = GetAggregate(id);
        var payload = _payloadGenerator.Generate(rtgsReturn);
        rtgsReturn.GeneratePayload(payload, DateTime.UtcNow);
        _repository.Save(rtgsReturn);
        return RtgsReturnResponse.From(rtgsReturn);
    }

    public RtgsReturnResponse MarkSent(Guid id)
    {
        var rtgsReturn = GetAggregate(id);
        rtgsReturn.MarkAsSent(DateTime.UtcNow);
        _repository.Save(rtgsReturn);
        return RtgsReturnResponse.From(rtgsReturn);
    }

    public RtgsReturnResponse Acknowledge(Guid id)
    {
        var rtgsReturn = GetAggregate(id);
        rtgsReturn.Acknowledge(DateTime.UtcNow);
        _repository.Save(rtgsReturn);
        return RtgsReturnResponse.From(rtgsReturn);
    }

    public RtgsReturnResponse Get(Guid id) => RtgsReturnResponse.From(GetAggregate(id));

    private RtgsReturn GetAggregate(Guid id) => _repository.GetById(id) ?? throw new ArgumentException($"RTGS Return not found: {id}");
}
